lib/datadog/tracing/contrib/action_pack/action_dispatch/instrumentation.rb

```diff
@@ -0,0 +1,93 @@
+# frozen_string_literal: true
+
+require_relative '../../../metadata/ext'
+
+module Datadog
+  module Tracing
+    module Contrib
+      module ActionPack
+        module ActionDispatch
+          # Instrumentation for ActionDispatch components
+          module Instrumentation
+            module_function
+
+            def format_http_route(http_route)
+              http_route.gsub(/\(.:format\)\z/, '')
+            end
+
+            # Instrumentation for ActionDispatch::Journey components
+            module Journey
+              # Instrumentation for ActionDispatch::Journey::Router
+              # for Rails versions older than 7.1
+              module Router
+                def find_routes(req)
+                  result = super
+
+                  return result unless Tracing.enabled?
+
+                  active_span = Tracing.active_span
+                  return result unless active_span
+
+                  begin
+                    # Journey::Router#find_routes retuns an array for each matching route.
+                    # This array is [match_data, path_parameters, route].
+                    # We need the route object, since it has a path with route specification.
+                    current_route = result.last&.last&.path&.spec
+                    return result unless current_route
+
+                    # When Rails is serving requests to Rails Engine routes, this function is called
+                    # twice: first time for the route on which the engine is mounted, and second
+                    # time for the internal engine route.
+                    last_route = active_span.get_tag(Tracing::Metadata::Ext::HTTP::TAG_ROUTE)
+
+                    active_span.set_tag(
+                      Tracing::Metadata::Ext::HTTP::TAG_ROUTE,
+                      Instrumentation.format_http_route(last_route.to_s + current_route.to_s)
+                    )
+                  rescue StandardError => e
+                    Datadog.logger.error(e.message)
+                  end
+
+                  result
+                end
+              end
+
+              # Since Rails 7.1 `Router#serve` adds `#route_uri_pattern` attribute to the request,
+              # and the `Router#find_routes` now takes a block as an argument to make the route computation lazy
+              # https://github.com/rails/rails/commit/35b280fcc2d5d474f9f2be3aca3ae7aa6bba66eb
+              module LazyRouter
+                def serve(req)
+                  response = super
```

Comment:
This makes the route resolved to *after* the request is processed by whatever is behind `super`, which means:

- route information is not available to other parts until after the fact
- for ASM it is not blockable based on route information (or rather it could block but  way too late when effects have been applied, possibly with disastrous consequences).

Here's the patch I had for Rails 7.1+ that gets the information at exactly the right time by hooking into the block†:

```diff
commit 1bc8c3fd94bd32abe79c0c29d9920bc193c7d9f2
Merge: 14b02858e f92f6b430
Author: Loic Nageleisen <loic.nageleisen@datadoghq.com>
Date:   Tue Mar 19 16:08:42 2024 +0100

    WIP on patch-sinatra-once: 14b02858e Move configuration of Rack integration within initial Datadog.configure

diff --cc lib/datadog/tracing/contrib/rails/patcher.rb
index e2d147cd0,e2d147cd0..07d4b4a7f
--- a/lib/datadog/tracing/contrib/rails/patcher.rb
+++ b/lib/datadog/tracing/contrib/rails/patcher.rb
@@@ -13,13 -13,13 +13,34 @@@ module Datado
        module Rails
          # Patcher to trace rails routing done by JourneyRouter
          module JourneyRouterPatch
--          def find_routes(*args)
--            result = super
--            integration_route = result.first[2].path.spec.to_s.gsub(/\(\.:format\)/, '') if result.any?
--            request_trace = Tracing.active_trace || TraceOperation.new
--            request_trace.set_tag(Tracing::Metadata::Ext::HTTP::TAG_ROUTE, integration_route)
--            request_trace.set_tag(Tracing::Metadata::Ext::HTTP::TAG_ROUTE_PATH, args.first.env['SCRIPT_NAME'])
--            result
++          def serve(req)
++            r = super
++          ensure
++            binding.irb
++          end
++
++          def find_routes(req, &block)
++            binding.irb
++            result = []
++            ret = super(req) do |match, parameters, route|
++              binding.irb
++              result << [match, parameters, route]
++              block.call(match, parameters, route)
++            ensure
++              binding.irb
++              p ret.nil?
++              p result.any?
++              p ret.nil? && result.any?
++              if ret.nil? && result.any?
++                binding.irb
++                integration_route = result.last[2].path.spec.to_s.gsub(/\(\.:format\)/, '')
++                request_trace = Tracing.active_trace || TraceOperation.new
++                request_trace.set_tag(Tracing::Metadata::Ext::HTTP::TAG_ROUTE, integration_route)
++                request_trace.set_tag(Tracing::Metadata::Ext::HTTP::TAG_ROUTE_PATH, req.env['SCRIPT_NAME'])
++              end
++            end
++          ensure
++            binding.irb
            end
          end
  
```

Note that the previous implementation which supported rails 7.0 and down worked out of a lucky `result.first` guess that is _usually_ correct (there's only one result) but not _guaranteed_ correct as multiple routes could theoretically be matched (the return value is an `Array`/`Enumerable`).

It would be interesting to have a test case where multiple routes are matched by the route resolver (IIRC in that case Rails takes the first one) to ensure that we have a correct understanding of this case + that our implementation is correct in this case as well.

† This kind of hooking will be made easier by https://github.com/DataDog/graft-rb including guaranteed ordering between `Tracing` and `AppSec`.

Reply:
Thank you for the review!

I changed patching of rails 7.1 [079872b](https://github.com/DataDog/dd-trace-rb/pull/3849/commits/079872b319334a9804ee42edb40318dd5ef11247) according to your suggestion.

And I also added [d5bbc94](https://github.com/DataDog/dd-trace-rb/pull/3849/commits/d5bbc947e8cad116c52cb4f94c62408e7cf64674) as you suggested.

Reply:
and thank you again for mentioning the importance of handling the route and script name separately, this helped to correctly handle nested Grape application routes